<!DOCTYPE html>
<html>
<head><script src="/livereload.js?port=1313&mindelay=10&v=2" data-no-instant defer></script>
	<meta charset="utf-8" />
	<meta http-equiv="X-UA-Compatible" content="IE=edge"><title>Handling the AggregateException - Jef Claes</title><meta name="viewport" content="width=device-width, initial-scale=1">
	<meta itemprop="name" content="Handling the AggregateException">
<meta itemprop="description" content="Last week I showed you how you can use the AggregateException to apply consistent exception handling in batch operations. You can find that post here.
Bart De Smet read that post and pointed out that I should check out the Handle method of the AggregateException.
The Handle methodAs found in the MSDN documentation.
Description
 Invokes a handler on each Exception contained by this AggregateException.
 Parameters
 System.Func&lt;Exception, Boolean&gt; predicate">
<meta itemprop="datePublished" content="2010-05-23T13:10:00&#43;02:00" />
<meta itemprop="dateModified" content="2010-05-23T13:10:00&#43;02:00" />
<meta itemprop="wordCount" content="366">



<meta itemprop="keywords" content="CodeSnippets,.NET,Tips," /><meta property="og:title" content="Handling the AggregateException" />
<meta property="og:description" content="Last week I showed you how you can use the AggregateException to apply consistent exception handling in batch operations. You can find that post here.
Bart De Smet read that post and pointed out that I should check out the Handle method of the AggregateException.
The Handle methodAs found in the MSDN documentation.
Description
 Invokes a handler on each Exception contained by this AggregateException.
 Parameters
 System.Func&lt;Exception, Boolean&gt; predicate" />
<meta property="og:type" content="article" />
<meta property="og:url" content="http://localhost:1313/2010/05/handling-the-aggregateexception/" />
<meta property="article:published_time" content="2010-05-23T13:10:00+02:00" />
<meta property="article:modified_time" content="2010-05-23T13:10:00+02:00" />
<meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="Handling the AggregateException"/>
<meta name="twitter:description" content="Last week I showed you how you can use the AggregateException to apply consistent exception handling in batch operations. You can find that post here.
Bart De Smet read that post and pointed out that I should check out the Handle method of the AggregateException.
The Handle methodAs found in the MSDN documentation.
Description
 Invokes a handler on each Exception contained by this AggregateException.
 Parameters
 System.Func&lt;Exception, Boolean&gt; predicate"/>
<link href='https://fonts.googleapis.com/css?family=Playfair+Display:700' rel='stylesheet' type='text/css'>
	<link rel="stylesheet" type="text/css" media="screen" href="http://localhost:1313/css/normalize.css" />
	<link rel="stylesheet" type="text/css" media="screen" href="http://localhost:1313/css/main.css" />

        <link id="dark-scheme" rel="stylesheet" type="text/css" href="http://localhost:1313/css/dark.css" />

	<script src="https://cdn.jsdelivr.net/npm/feather-icons/dist/feather.min.js"></script>
		<script src="http://localhost:1313/js/main.js"></script>
</head>

<body>
	<div class="container wrapper">
		<div class="header">
	
		<div class="avatar">
			<a href="http://localhost:1313/">
				<img src="https://image.freepik.com/free-vector/young-man-head-with-beard-avatar-character_24877-36786.jpg" alt="Jef Claes" />
			</a>
		</div>
	
	<h1 class="site-title"><a href="http://localhost:1313/">Jef Claes</a></h1>
	<div class="site-description"><p>On software and life</p><nav class="nav social">
			<ul class="flat"><li><a href="https://github.com/jefclaes" title="Github"><i data-feather="github"></i></a></li><li><a href="/index.xml" title="RSS"><i data-feather="rss"></i></a></li></ul>
		</nav></div>

	<nav class="nav">
		<ul class="flat">
			
			<li>
				<a href="/">Home</a>
			</li>
			
			<li>
				<a href="/post">All posts</a>
			</li>
			
			<li>
				<a href="/about">About</a>
			</li>
			
			<li>
				<a href="/tags">Tags</a>
			</li>
			
		</ul>
	</nav>
</div>


		<div class="post">
			<div class="post-header">
				
					<div class="meta">
						<div class="date">
							<span class="day">23</span>
							<span class="rest">May 2010</span>
						</div>
					</div>
				
				<div class="matter">
					<h1 class="title">Handling the AggregateException</h1>
				</div>
			</div>
					
			<div class="markdown">
				<p>Last week I showed you how you can use the AggregateException to apply
consistent exception handling in batch operations. You can find that
post
<a href="http://jclaes.blogspot.com/2010/05/exception-handling-in-batch-operations.html">here</a>.</p>
<p><a href="http://bartdesmet.net/blogs/bart/">Bart De Smet</a> read that post and
pointed out that I should check out the Handle method of the
AggregateException.</p>
<p><!-- raw HTML omitted -->The Handle method<!-- raw HTML omitted --></p>
<p>As found in the <a href="http://msdn.microsoft.com/en-us/library/system.aggregateexception.handle.aspx">MSDN
documentation</a>.</p>
<p>Description</p>
<blockquote>
<p>Invokes a handler on each Exception contained by this
AggregateException.</p>
</blockquote>
<p>Parameters</p>
<blockquote>
<p>System.Func&lt;Exception, Boolean&gt; predicate<br>
The predicate to execute for each exception. The predicate accepts as
an argument the Exception to be processed and returns a Boolean to
indicate whether the exception was handled.</p>
</blockquote>
<p>Remarks</p>
<blockquote>
<p>Each invocation of the predicate returns true or false to indicate
whether the Exception was handled. After all invocations, if any
exceptions went unhandled, all unhandled exceptions will be put into a
new AggregateException which will be thrown. Otherwise, the Handle
method simply returns. If any invocations of the predicate throws an
exception, it will halt the processing of any more exceptions and
immediately propagate the thrown exception as-is.</p>
</blockquote>
<p><!-- raw HTML omitted -->In practice<!-- raw HTML omitted --></p>
<p>I refactored the example in <a href="http://jclaes.blogspot.com/2010/05/exception-handling-in-batch-operations.html">my previous
post</a>
to make use of the Handle method.</p>
<pre><code>   1:  static void Main(string[] args)

   2:  {

   3:       try

   4:       {

   5:            ExecuteBatch();

   6:       }

   7:       catch (AggregateException aggEx)

   8:       {

   9:            aggEx.Handle(HandleBatchExceptions);

  10:       }

  11:  Â 

  12:       Console.ReadLine();

  13:  }
</code></pre>
<p>I&rsquo;m passing a <a href="http://msdn.microsoft.com/en-us/library/bb549151.aspx">Func&lt;T, TResult&gt;
delegate</a> to the
Handle method. In this delegate I decide whether I&rsquo;m handling the
exception or not. If I handle the exception, I return true, else I
return false.</p>
<pre><code>   1:  private static bool HandleBatchExceptions(Exception exceptionToHandle)

   2:  {

   3:       if (exceptionToHandle is ArgumentNullException)

   4:       {

   5:            //I'm handling the ArgumentNullException.

   6:            Console.WriteLine(&quot;Handling the ArgumentNullException.&quot;);

   7:            //I handled this Exception, return true.

   8:            return true;

   9:       }

  10:       else

  11:       {

  12:            //I'm only handling ArgumentNullExceptions.

  13:            Console.WriteLine(string.Format(&quot;I'm not handling the {0}.&quot;, exceptionToHandle.GetType()));

  14:            //I didn't handle this Exception, return false.

  15:            return false;

  16:       }          

  17:  }
</code></pre>
<p>When we run this example a new AggregateException is thrown with the
exceptions I didn&rsquo;t handle.</p>
<p><a href="/post/images/2010-05-23-handling-the-aggregateexception-ConsoleOut.bmp"><img src="/post/images/thumbnails/2010-05-23-handling-the-aggregateexception-ConsoleOut.bmp" alt=""></a><br>
<a href="/post/images/2010-05-23-handling-the-aggregateexception-Rethrown.bmp"><img src="/post/images/thumbnails/2010-05-23-handling-the-aggregateexception-Rethrown.bmp" alt=""></a></p>
<p><!-- raw HTML omitted -->Conclusion<!-- raw HTML omitted --></p>
<p>Make use of the Handle method to run over each InnerException and decide
which exception you want to handle or not. The exceptions you didn&rsquo;t
handle are automatically wrapped in a new AggregateException which gets
rethrown.</p>

			</div>

			<div class="tags">
				
					
						<ul class="flat">
							
							<li><a href="/tags/codesnippets">CodeSnippets</a></li>
							
							<li><a href="/tags/.net">.NET</a></li>
							
							<li><a href="/tags/tips">Tips</a></li>
							
						</ul>
					
				
			</div></div>
	</div>
	<div class="footer wrapper">
	<nav class="nav">
		<div>2010  <a href="https://github.com/knadh/hugo-ink">Ink</a> theme on <a href="https://gohugo.io">Hugo</a></div>
	</nav>
</div>

<script>feather.replace()</script>
</body>
</html>
